java/arcs/core/analysis/Deduction.kt

```diff
@@ -0,0 +1,179 @@
+
+  /** [Deduction]s that have a simple path (a list of [Identifier]s). */
+  interface Pathlike {
+    val path: List<Identifier>
+    fun mergeTop(other: Pathlike): Deduction
+  }
+
+  /** A representation of a field path in a Paxel [Expression]. */
+  data class Path(override val path: List<Identifier> = emptyList()) : Deduction(), Pathlike {
+
+    constructor(vararg paths: Identifier) : this(paths.toList())
+
+    /** Base-case: Returns self as underlying path.*/
+    override fun getPath(): Path = this
+
+    /** Return [Path] with all alias substitutions applied. */
+    override fun substitute(aliases: Scope): Path =
+      Path(
+        path.flatMap { identifier ->
+          aliases.associations.getOrDefault(identifier, Paths(Path(identifier))).getPath().path
+        }
+      )
+
+    /** Append more [Identifier]s to the [Path] */
+    override fun mergeTop(other: Pathlike) = Path(path + other.path)
+
+    /** Union of a [Path] and a [Deduction]. */
+    override fun plus(other: Deduction): Deduction = when (other) {
+      is Path -> Paths(this) + Paths(other)
+      is Paths -> Paths(this) + other
+      is Scope -> Paths(this) + Paths(other)
+      else -> this + other
+    }
+  }
+
+  /** A collection of field [Path]s. */
+  data class Paths(val paths: List<Deduction> = emptyList()) : Deduction() {
+
+    constructor(vararg paths: Deduction) : this(paths.toList())
+
+    constructor(vararg paths: List<Identifier>) : this(paths.map { Path(it) })
+
+    /** Union of a [Paths] and another [Deduction]. */
+    override fun plus(other: Deduction): Deduction = when (other) {
+      is Path -> this + Paths(other)
+      is Paths -> Paths(paths + other.paths)
+      is Scope -> this + Paths(other)
+      else -> this + other
+    }
+
+    /** Returns the first [Path] in the collection, or an empty [Path]. */
+    override fun getPath(): Path = if (paths.isNotEmpty()) paths.first().getPath() else Path()
+
+    /** Substitute all aliases as a new [Paths] object. */
+    override fun substitute(aliases: Scope) =
+      Paths(paths.map { path -> path.substitute(aliases) })
+  }
+
+  /** Associates an [Identifier] with a group of [Deduction]s. */
+  data class Scope(
+    val associations: Map<Identifier, Deduction> = emptyMap()
+  ) : Deduction() {
+
+    constructor(vararg pairs: Pair<Identifier, Deduction>) : this(pairs.toMap())
+
+    /** Returns a new [Scope] with only the queried association. */
+    fun lookup(key: Identifier): Scope {
+      require(key in associations) {
+        "Scope does not associate anything with '$key'."
+      }
+      return Scope(key to associations.getOrDefault(key, Paths()))
+    }
+
+    /** Union of a [Scope] and another [Deduction]. */
+    override fun plus(other: Deduction): Deduction = when (other) {
+      is Path -> Paths(this) + Paths(other)
+      is Paths -> Paths(this) + other
+      is Scope -> Scope(this.associations + other.associations)
```

Comment:
Here you will drop some association possibly - this is dangerous. You should leave this unimplemented for now or handle it in this PR.

e.g.
```
union(
  from f in foo select new X {x: f.a},
  from f in foo select new X {x: f.b}
).x
```

A conservative derivation would be Derive(foo.a, foo.b)

When you evaluate the above you will need to do the merging of scope obtained from 2 arguments of the union, and the result be that for every identifier available in both scopes you produce merge of their deductions.

Reply:
I can see how this operation could be dangerous. However, I am fairly sure that with the current implementation, this union example will produce the correct derivation. This is because the `select` expression is evaluated before the `union` and it will apply the substitution.

Reply:
On second look: I see -- `x` is the identifier that will cause the collision.